File: p1/src/cpu.py
# Michael LeMay, 1/19/19

import cache


class Cpu:


    def __init__(self, numSets, numBlocksPerSet,
                 replacementPolicy, writePolicy, blockSize, ramSize):
        """Creates the cpu and its parameters. Creates the cache.
        Cpu also counts the number of instructions it receives.
        """
        self.numSets = numSets
        self.numBlocksPerSet = numBlocksPerSet
        self.ramSize = ramSize
        self.replacementPolicy = replacementPolicy
        self.writePolicy = writePolicy
        self.blockSize = blockSize

        self.cache = cache.Cache(numSets,
                                 numBlocksPerSet,
                                 replacementPolicy,
                                 writePolicy,
                                 blockSize,
                                 ramSize)

        self.instructionCount = 0


    def loadDouble(self, addr):
        """Increments instruction count and tells cache to return a value at
        an address.
        """
        self.instructionCount += 1
        return self.cache.getDouble(addr)


    def storeDouble(self, addr, value):
        """Increments instruction count and tells cache to set an address
        to a given value.
        """
        self.instructionCount += 1
        self.cache.setDouble(addr, value)


    def addDouble(self, v1, v2):
        """Adds two doubles together and increments instruction count"""
        self.instructionCount += 1
        return v1 + v2


    def multDouble(self, v1, v2):
        """Multiplies two doubles and increments instruction count"""
        self.instructionCount += 1
        return v1 * v2


    def cpuStats(self):
        """Gets stats from cache and prints out the hits, misses, miss rate, and
        instruction count.
        """
        ret = {}
        ret["instructionCount"] = self.instructionCount
        cacheStats = self.cache.getCacheCountingInfo()
        ret["readHits"] = cacheStats["readHits"]
        ret["readMisses"] = cacheStats["readMisses"]
        ret["writeHits"] = cacheStats["writeHits"]
        ret["writeMisses"] = cacheStats["writeMisses"]
        print("Instructions Run:", ret["instructionCount"])
        print("Read hits:       ", cacheStats["readHits"])
        print("Read misses:     ", cacheStats["readMisses"])
        print("Write hits:      ", cacheStats["writeHits"])
        print("Write misses:    ", cacheStats["writeMisses"])
        if (ret["readHits"] + cacheStats["readMisses"]) != 0:
            print("Read miss rate:  ", cacheStats["readMisses"] /
                  (cacheStats["readHits"] + cacheStats["readMisses"]))
        if (ret["writeHits"] + cacheStats["writeMisses"]) != 0:
            print("Write miss rate: ", cacheStats["writeMisses"] /
                  (cacheStats["writeHits"] + cacheStats["writeMisses"]))

File: p1/src_c++/cpu.cpp
/* Michael LeMay, 1/13/19 */

#include "cpu.h"
#include "address.h"
#include "cache.h"
#include <iostream>
#include <string>


/* CPU implementation */

/*
 * Constructor for CPU. Initiates ram and cache.
 */
Cpu::Cpu(int newNumSets, int newNumBlocks,
         std::string theReplacementPolicy,
	 std::string theWritePolicy) {
  /*Load in private variables*/
  numSets = newNumSets;
  numBlocks = newNumBlocks;
  replacementPolicy = theReplacementPolicy;
  writePolicy = theWritePolicy;

  /*Create the associated cache object the cpu needs*/
  cache = Cache(newNumSets, newNumBlocks, replacementPolicy, writePolicy);
}

/*
 * This gets a double from ram or cache based on the address
 */
double Cpu::loadDouble(Address addr) {
  /*Ask for value from Cache at addr*/
  double doubleAtAddr = cache.getDouble(addr);
  /*Return the desired double*/
  return doubleAtAddr;
}


/*
 * Loads a double into cache, or ram
 */
void Cpu::storeDouble(Address addr, double value) {
  cache.setDouble(addr, value);
}


/*
 * Add two doubles together
 */
double Cpu::addDouble(double value1, double value2) {
  return value1 + value2;
}


/*
 * Multiply two doubles together
 */
double Cpu::multDouble(double value1, double value2) {
  return value1 * value2;
}
